client/components/dashboard/dashboard-service.js

```diff
@@ -367,56 +371,95 @@ DashboardService.prototype.initializeParams_ = function() {
  */
 DashboardService.prototype.selectWidget = function(
     widget, container, opt_supressStateChange) {
+  let currentContainer = this.explorerStateService_.containers.selected;
   let currentWidget = this.explorerStateService_.widgets.selected;
 
-  if (currentWidget) {
-    currentWidget.state().selected = false;
+  if (currentWidget !== widget) {
+    currentWidget && (currentWidget.state().selected = false);
+    widget && (widget.state().selected = true);
   }
 
-  let currentContainer = this.explorerStateService_.containers.selected;
-
-  if (currentContainer && currentContainer !== container) {
-    currentContainer.state().selected = false;
+  if (currentContainer !== container) {
+    currentContainer && (currentContainer.state().selected = false);
+    container && (container.state().selected = true);
   }
 
-  if (container && currentContainer !== container) {
-    container.state().selected = true;
+  if (!opt_supressStateChange) {
+    params = {widget: undefined, container: undefined};
+
+    if (widget) { params.widget = widget.model.id; }
+    if (container) { params.container = container.model.id };
+
+    this.$state_.go('explorer-dashboard-edit', params);
   }
 
   if (widget) {
-    widget.state().selected = true;
-
-    if (!(this.sidebarTabService_.selectedTab &&
-          this.sidebarTabService_.selectedTab.requireWidget)) {
+    if (this.sidebarTabService_.selectedTab &&
+        !this.sidebarTabService_.selectedTab.requireWidget) {
       this.sidebarTabService_.selectTab(
           this.sidebarTabService_.getFirstWidgetTab());
     }
+
+    this.timeout_(() => {
+      this.scrollWidgetIntoView(widget);
+    });
+  } else if (container) {
+    if (this.sidebarTabService_.selectedTab &&
+        !this.sidebarTabService_.selectedTab.requireContainer) {
+      this.sidebarTabService_.selectTab(
+          this.sidebarTabService_.getFirstContainerTab());
+    }
+
+    this.timeout_(() => {
+      this.scrollContainerIntoView(container);
+    });
+  } else {
+    this.timeout_(() => {      
+      if (this.sidebarTabService_.selectedTab &&
+          !this.sidebarTabService_.isTabVisible(this.sidebarTabService_.selectedTab)) {
+        this.sidebarTabService_.selectTab(
+            this.sidebarTabService_.getFirstTab());
+      }
+    });
   }
+};
 
-  this.timeout_(() => {
-    this.scrollWidgetIntoView(widget);
-  });
 
-  if (!opt_supressStateChange) {
-    params = {widget: undefined, container: undefined};
 
-    if (widget) { params.widget = widget.model.id; }
-    if (container) { params.container = container.model.id };
+/**
+ * Scrolls the specified content element (typically a widget or container) into view.
+ * @param {!angular.Element} targetElement The element to scroll into view.
+ */
+DashboardService.prototype.scrollPageElementIntoView = function(targetElement) {
+  let contentElement = angular.element(document.getElementsByClassName('pk-page-content'));
```

Comment:
We're using CSS classes to identify elements in general, for two reasons:
* Consistency - ID's are expected to be unique to the document; when used in ng-repeat sections, it will violate this rule.  Class names, on the other hand, can be used as many times as needed.  The ability to search for elements relative to an ancestor (as getElementsByClassName supports) gives you a means to find elements with duplicate "id's" when they're based on CSS classes.
* Extensibility - Future capabilities for third-party content add the possibility of an id collision, even if we worked around/ignored the above consistency issue.